GPII-2521: Merging of Morpher and Simplification into upstream

JavierJF opened the pull request:
This pull is blocked by [this other pull request.](https://github.com/GPII/universal/pull/554)

stegru committed:
GPII-1716: Added the ability to set fWinIni SPI parameter from the solution registry.

stegru committed:
GPII-1716: Added font-size support, via SPI_SETNONCLIENTMETRICS and SPI_SETICONTITLELOGFONT

gloob committed:
Pointing to universal#demo-mainstream.

gloob committed:
GPII-2376 : Adding needed installation bits for W10 Simplification.

gloob committed:
GPII-2376 : Returning 0 at the end of the script if everything went
fine.

gloob committed:
GPII-2376 : Adding --yes and --force to the chocolatey installation.

JavierJF committed:
Fixed bad option parameter and silenced command

JavierJF committed:
Added necessary scripts to synchronize start/stop lifecycle handlers

JavierJF committed:
Unified provisioning scripts for morpher

JavierJF committed:
Removed unnecessary lines

JavierJF committed:
Removed unnecessary comment

gloob committed:
Pointing to universal instead of windows.

gloob committed:
Moving all the demo-data to the gpii-app repository instead of the
Windows. All the action will be triggered by the gpii-app repository.

gloob committed:
Chaging protocol of universal from ssh to https.

gloob committed:
Use relative path using script main dir as a root path.

gloob committed:
Changed mainDir inferer system based on the initiator script.

gloob committed:
Morpher setup script wasn't updating the GPII-Demo directory.

stegru committed:
GPII-1716: Updated reference to the working universal branch.

stegru committed:
GPII-1716: Minor quality improvements.

stegru committed:
GPII-2495: Made SPI tests work again.

stegru committed:
GPII-2212: Added restart/shutdown applications

stegru committed:
GPII-2212: Pointing universal to the branch that works with this.

stegru committed:
Merge remote-tracking branch 'upstream/master' into GPII-2212

# Conflicts:
#	package.json

JavierJF committed:
GPII-2521: Functions for ensure the state of a process after command exec

JavierJF committed:
GPII-2521: Test for new process handling functions

JavierJF committed:
GPII-2521: Modifications to SpiSettingsHandler and WindowUtilities for handling wstrings and fWinIni flags

JavierJF committed:
GPII-2521: Added test for new supported payload for setting handler

JavierJF committed:
GPII-2521: New package needed to work with wchar in node ffi

JavierJF committed:
GPII-2521: Added image as data payload for setting handler tests

stegru committed:
Merge remote-tracking branch 'upstream/master' into GPII-2212

# Conflicts:
#	gpii/node_modules/WindowsUtilities/WindowsUtilities.js
#	gpii/node_modules/processHandling/processHandling.js
#	gpii/node_modules/processHandling/test/testProcessHandling.js
#	package.json

JavierJF committed:
Merge remote-tracking branch 'upstream/master' into GPII-2521

JavierJF committed:
GPII-2521: Fixed function name typo

JavierJF committed:
GPII-2521: Added missing payload for wallpaper test

JavierJF committed:
GPII-2521: Created new folder for scripts making env changes

Added script for creating a folder for GPII data

JavierJF committed:
GPII-2521: Created new folder for chocolatey pkgs

The folder chocopkgs is pretended to hold the scripts for the installation of
packages that need more logic that a simply package installation.

JavierJF committed:
GPII-2521: Added scripts and menu layout for immersive shell and custom start menu

JavierJF committed:
GPII-2521: Added custom windows logo and wallpapers for morpher

JavierJF committed:
GPII-2521: Updated package.json to point the dependent gpii-universal branch

JavierJF committed:
Merge remote-tracking branch 'upstream/master' into GPII-2521

JavierJF committed:
GPII-2521: Fixed typo in repository reference

stegru committed:
Merge remote-tracking branch 'upstream/master' into GPII-2212

# Conflicts:
#	gpii/node_modules/WindowsUtilities/WindowsUtilities.js
#	package.json

JavierJF committed:
GPII-2521: Added extra verification for SpiSettingHandler: SPI_SETDESKWALLPAPER action

amb26 commented:
Cheers, @JavierJF - I've updated the linting configuration in master to match that in our other projects. Please could you regenerate this pull against master since it has picked up some changes from the process reporter work, and then make sure it passes linting - thanks!

JavierJF committed:
Merge branch 'master' of https://github.com/GPII/windows into GPII-2521

JavierJF committed:
Improved script soundness with bitness check and better error reporting

JavierJF committed:
GPII-2521: Added check for script being launch in 64bit mode in 64bit systems

JavierJF committed:
Merge branch 'GPII-2521' of https://github.com/JavierJF/windows into GPII-2521

JavierJF committed:
GPII-2521: Fixed all errors reported by eslint

amb26 commented on gpii/node_modules/spiSettingsHandler/src/SpiSettingsHandler.js:
```diff
@@ -260,6 +315,96 @@ gpii.windows.spi.populateResults = function (payload, isNewValue, isGetting, res
+ * @param {*} s String to count its words.
+ * @return The number of words in the string parameter.
+ */
+function countWords(s) {
```
Interesting algorithm. Given it handles so many cases, we should provide tests for them - or else simplify it so that it handles a more restricted range of possibilities. However, I'm concerned by the asymmetry between "countWord" and "splitInWord". Really, they should operate the same algorithm or at least algorithms which are complementary.
Whatever we do in this area should also be exported as a public function (which you would need to do anyway in order to write tests - such wonders of test-oriented development!)

JavierJF replied:
Nice! I will start exporting it as a public function, and writing some tests for it. Then we could talk about the cases in which we could reuse it.

amb26 commented on gpii/node_modules/spiSettingsHandler/src/SpiSettingsHandler.js:
```diff
@@ -260,6 +315,96 @@ gpii.windows.spi.populateResults = function (payload, isNewValue, isGetting, res
+/**
+ * @param {String} s Returns true if value is a string, false otherwhise.
+ */
+function isString(s) {
```
I am baffled that this function is necessary - what are these strings which fail the typeof check and yet have a constructor which is a string?

JavierJF replied:
That was just for checking strings that were constructed using 'new operator', then 'typeof check' will fail for them, since it will return 'object'.

amb26 replied:
The construction of strings using the new operator is such a crazy practice that there is even an eslint rule to enforce against it https://eslint.org/docs/rules/no-new-wrappers - I notice that this is not in our community rules but I will propose having it added. You may safely assume that noone in our community has done this and simply use typeof (x) === "string"

amb26 replied:
Note the following answer: https://stackoverflow.com/questions/5750656/whats-the-point-of-new-stringx-in-javascript - "No native JavaScript object, major library or DOM method that returns a string will return a String object rather than a string value."

JavierJF replied:
Didn't knew about the eslint rule against it, or the global community agreement against it, if it's in our community guidelines I agree also that we could asume that will never happen. Removing it!

amb26 commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -296,6 +297,120 @@ gpii.windows.closeProcessByName = function (filename, options) {
+ * @return {promise}
+ */
+gpii.windows.closeAfterExitCommand = function (command, args, processToWait, processToClose, options) {
+    var defaultOptions = {
```
Factor this repeated definition into a public one, perhaps even a non-component defaults block

JavierJF replied:
Okay, then I will refactor both of then into closeAfterCommand, and change the flow with the payload. What should I do with the 'non-component default block'.

amb26 replied:
You could just say fluid.defaults("gpii.windows.closeAfterCommand", {
   ....
}
and then retrieve them again using the same call. This doesn't offer particularly many benefits other than allowing you to use the same global name for the function as well as for the options.

You can see this pattern used in the framework here: https://github.com/fluid-project/infusion/blob/master/src/framework/core/js/FluidIoC.js#L2305

As a funny historical note, this is all that Infusion defaults ever were, when they were first invented (around 2008)

amb26 commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -296,6 +297,120 @@ gpii.windows.closeProcessByName = function (filename, options) {
 };
 
+/**
+ * @param command The command to be executed for closing an application.
```
These two functions are nearly identical and should be factored together. Also, there should be a clear explanation of the purpose and workflow of the function.

amb26 commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -296,6 +297,120 @@ gpii.windows.closeProcessByName = function (filename, options) {
+
+    // Linter (Bug) Workaround: Linter fails with arrow style lambda with multiple parameters.
+    closeCommand.on('exit', function (code, signal) {
+        // Linter (Bug) Workaround: Linter fails with arrow style lambda without parameters.
```
It is not a bug, it is simply that such ES6 neologisms are not part of our project standard :)
I think you will find that when you factor these functions better, the desirability of them will fade away a bit ...

JavierJF replied:
Ups... My bad, I will pass again the linter and change the comments about the bug.

amb26 commented:
I think we will have to leave the wallpapers out of the repo since they don't meet our licence conditions. Let's just go with solid colour backgrounds, or open licenced placeholders, for now, and if this is queried we can start the wider licencing discussion.

amb26 commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -296,6 +297,120 @@ gpii.windows.closeProcessByName = function (filename, options) {
+                    } else {
+                        promise.resolve(true);
+                    }
+                }, (err) => {
```
This ES6 neologism also fails linting (not a bug : P )

amb26 commented:
For me the Unit tests for this branch fail with the following: 

07:37:32.967:  {
    "module": "GPII",
    "event": "Error.Fail",
    "level": "FAIL",
    "data": {
        "error": {
            "message": "gpii.windows.spi.applySettings: Invalid PvParam payload '\u0000\u0000'",
            "stack": "Error\n    at gpii.eventLog.logError (\\\\vboxsrv\\vagrant\\node_modules\\universal\\gpii\\node_modules\\eventLog\\src\\eventLog.js:187:28)\n    at invokeInvoker [as logError] (\\\\vboxsrv\\vagrant\\node_modules\\infusion\\src\\framework\\core\\js\\FluidIoC.js:1736:29)\n    at Function.gpii.eventLog.gotError (\\\\vboxsrv\\vagrant\\node_modules\\universal\\gpii\\node_modules\\eventLog\\src\\eventLog.js:208:21)\n    at \\\\vboxsrv\\vagrant\\node_modules\\universal\\gpii\\node_modules\\eventLog\\src\\eventLog.js:197:19\n    at Object.fire (\\\\vboxsrv\\vagrant\\node_modules\\infusion\\src\\framework\\core\\js\\Fluid.js:1533:40)\n    at Object.fluid.fail (\\\\vboxsrv\\vagrant\\node_modules\\infusion\\src\\framework\\core\\js\\Fluid.js:213:32)\n    at Object.gpii.windows.spi.applySettings (\\\\ .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
07:37:32.987:  jq: FAIL: Module "gpii.tests.acceptance.windows.builtIn.config tests" Test name "Testing os_win7 using default matchmaker" - Message: Assertion failure (see console.log for expanded message): gpii.windows.spi.applySettings: Invalid PvParam payload '  '
07:37:32.987:  jq: Source:     at jqUnit.nodeFailureHandler (\\vboxsrv\vagrant\node_modules\node-jqunit\lib\jqUnit-node.js:34:15)
    at Object.fire (\\vboxsrv\vagrant\node_modules\infusion\src\framework\core\js\Fluid.js:1533:40)
    at Object.fluid.fail (\\vboxsrv\vagrant\node_modules\infusion\src\framework\core\js\Fluid.js:213:32)
    at Object.gpii.windows.spi.applySettings (\\vboxsrv\vagrant\gpii\node_modules\spiSettingsHandler\src\SpiSettingsHandler.js:245:19)
    at gpii.windows.spi.setImpl (\\vboxsrv\vagrant\gpii\node_modules\spiSettingsHandler\src\SpiSettingsHandler.js:514:22)
    at \\vboxsrv\vagrant\node_modules\universal\gpii\node_modules\settingsHandlers\src\settingsHandlerUtilities.js:150:21
    at Object.<anonymous> (\\vboxsrv\vagrant\node_modules\universal\gpii\node_modules\settingsHandlers\src\settingsHandlerUtilities.js:78:20)
    at transformInternal (\\vboxsrv\vagrant\node_modules\infusion\src\framework\core\js\Fluid.js:469:34)
    at Object.fluid.transform (\\vboxsrv\vagrant\node_modules\infusion\src\framewo .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]

JavierJF committed:
GPII-2521: Updated start menu layout for Simi due to changes in start menu icon policies

JavierJF committed:
GPII-2521: Fixed failing test due to race condition with ClassicShell exit command and daemon

JavierJF committed:
GPII-2521: Refactored functions closeAfterExitCommand and closeAfterOpenCommand into closeAfterCommand

JavierJF committed:
GPII-2521: Fixed typo in process restarting function

amb26 commented:
On my system the increased timeout is still insufficient - 

```
07:03:32.745:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:33.603:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:34.340:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:35.512:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:36.729:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:37.463:  {
    "module": "metrics",
    "event": "app-active",
    "level": "INFO",
    "data": {
        "exe": "C:\\Windows\\System32\\cmd.exe",
        "duration": 5
    }
}
07:03:37.463:  {
    "module": "metrics",
    "event": "app-launch",
    "level": "INFO",
    "data": {
        "exe": "C:\\Windows\\explorer.exe"
    }
}
07:03:37.666:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:37.745:  Test case listener has not responded after 5000ms - at sequence pos 11 of 19 sequence element {
    "event": "{exec}.events.onExecExit",
    "listener": "gpii.test.onExecExit"
} of fixture Testing ClassicShell
07:03:38.338:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:38.979:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:39.621:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:40.247:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:40.873:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:41.527:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:42.166:  Executing: tasklist /fi "STATUS eq RUNNING" /FI "IMAGENAME eq ClassicShell.exe" | find /I "ClassicShell.exe" /C
07:03:42.308:  jq: FAIL: Module "gpii.tests.acceptance.windows.classicShell.config tests" Test name "Testing ClassicShell" - Message: Checking the process with command: [object Object] - at sequence position 11 of 19
07:03:42.308:  jq: Source:     at pok (\\vboxsrv\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.assertTrue (\\vboxsrv\vagrant\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:145:13)
    at gpii.test.onExecExit (\\vboxsrv\vagrant\node_modules\universal\gpii\node_modules\testing\src\Testing.js:182:12)
    at wrapped (\\vboxsrv\vagrant\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:320:26)
    at Object.fire (\\vboxsrv\vagrant\node_modules\infusion\src\framework\core\js\Fluid.js:1533:40)
    at \\vboxsrv\vagrant\node_modules\universal\gpii\node_modules\testing\src\Acceptance.js:55:40
    at ChildProcess.exithandler (child_process.js:213:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:877:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
```

amb26 commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -296,6 +297,142 @@ gpii.windows.closeProcessByName = function (filename, options) {
+    var b64Cmd = process.env.windir + "\\sysnative\\cmd.exe";
+    child_process.execFile(b64Cmd, ["/k", processToRestart]);
+
+    gpii.windows.waitForProcessStart(processToRestart, {timeout: 1000, pollDelay:100})
```
These changes do not pass linting

amb26 commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -296,6 +297,142 @@ gpii.windows.closeProcessByName = function (filename, options) {
+        fluid.log("Waiting for process start: ");
+        gpii.windows.waitForProcessStart(processToWait, {timeout: 1000, pollDelay: 100})
+            .then(function () {
+                gpii.windows.closeProcessByName(processToClose, {cleanOnly: false, timeout: 150})
```
These promise handlers are still 100% duplicated

amb26 commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -296,6 +297,142 @@ gpii.windows.closeProcessByName = function (filename, options) {
     return promiseTogo;
 };
 
+gpii.windows._restartProcess = function(promise, processToRestart, retries) {
```
The "underscore" convention isn't used within our projects - instead, please supply some doc comments for these functions

amb26 commented:
This failure is in the real system, with what looks like the 15 second timeout. 
Also, we need to remove the copyrighted Microsoft wallpapers and images from this pull before it can be merged.

stegru committed:
Merge remote-tracking branch 'upstream/master' into GPII-2212

JavierJF committed:
Merge branch 'GPII-2212' of https://github.com/stegru/windows into GPII-2521

JavierJF committed:
GPII-2521: Refactored function for nicely closing processes after calling a command

stegru committed:
GPII-2212: MAde the teardowns global a member of the test module

JavierJF committed:
GPII-2521: Removed unused require statement

JavierJF committed:
Merge branch 'GPII-2212' of https://github.com/stegru/windows into GPII-2521

JavierJF committed:
GPII-2212: Fixed tests execution

stegru committed:
GPII-2212

Completed array to string change.

stegru committed:
Merge pull request #1 from JavierJF/GPII-2212

GPII-2212: Fixed tests execution

stegru committed:
GPII-2212: Fixed incorrect teardown

JavierJF committed:
Merge branch 'GPII-2212' of https://github.com/stegru/windows into GPII-2521

JavierJF committed:
GPII-2521: Refactored functions closeAppWithMinForce and closeAfterCommand

Changes in closeAppWithMinForce:
    + Function now waits for process termination after each attempt of
      closing the application.
    + Improved documentation.

Changes in closeAfterCommand:
    + Function now use internal callback for retrying the command exec.
    + Improved documentation.
    + Fixed parameters checks.

JavierJF committed:
GPII-2521: Added tests for functions 'closeAfterCommand' and 'closeAppWithMinForce'

stegru commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -487,6 +350,460 @@ gpii.windows.getProcessPath = function (pid, processHandle) {
+};
+
+/**
+ * Convenience function for trying to close an application by name, with the minimal necessary level of force.
```
`gpii.windows.closeApplications` already does this - if you pass "high", it will try the low, then medium if low didn't work, then high.

Just passing procPids to closeApplications should be enough.

JavierJF replied:
I realized, but it doesn't have any kind of wait for the process being stopped after any of the different force level calls, so I assumed that it simply 'tries more things' depending on the level of force.

stegru replied:
ok, I understand.

`RmShutdown.async` (line 498) calls back when the process has been stopped, and the "high" method is invoked after a 3 second delay (or options.quitDelay) after "medium".

I think calling this function with "low" is enough, then call `gpii.windows.closeProcessByName` when it resolves.

JavierJF replied:
Ok, so you propose to change the implementation to:

 *Call 'closeApplications' with "low", if that isn't enough, call 'closeProcessByName'*

Am I wrong?

stegru replied:
Yes. The purpose of closeApplications was to simulate a logout/in.

I'll be changing this stuff soon anyway, with the refactoring of GPII-2578.

JavierJF replied:
I have been seen not expected behavior when I have replaced this implementation with the suggested one using `closeProcessByName`, I think the reason is that `closeProcessByName` doesn't restart the process as the other function is supposed to do. I believe that the combination of registering the application for restart, and calling 'closeProcessByName' leads to process killing, and the restart never happens.

stegru commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -487,6 +350,460 @@ gpii.windows.getProcessPath = function (pid, processHandle) {
+
+/**
+ * Close a running application after a command has been executed.
+ *
```
It took me some time to understand what this actually does. It's not obvious (to me) what it does based on the documentation and signature.

I suggest documenting the primary use of this function (restarts explorer after starting/stopping classic start menu), and renaming it.

My understanding of it is this (ignoring what processA does):
- Starts processA (command)
- Waits for processB (processToWait) to start/stop
- Closes processC (processToClose)

In the only use case of this function, processA and B are both the same executable, so perhaps default `processToWait` to `command` (without the path), and move it into `options`.

This will simplify the function without changing the way it works:

- Starts processA, and wait for it to start/stop.
- Close processC.

Now the function can look like `startAndClose(command, processToClose, options)`, and anything that alters the meaning of the method name is hidden in options.

JavierJF replied:
I agree it looks more clear with that adjustments, and indeed the target was 'processA' and 'processB' being the same. I will reformat this function with your suggestion.

stegru replied:
Yes, find a better name.

I'm not convinced this stuff belongs in a single function, but it works around the fact that the solutions registry doesn't "chain" multiple commands. Maybe the work @kaspermarkus is doing fixes this.

JavierJF replied:
I'm not also, this would sound more natural *as a composition of functions*, the problem as you pointed out, is that we need a entre point for the solution registry.  So I will try to find the *less ugly | more descriptive name* for this.

stegru commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -487,6 +350,460 @@ gpii.windows.getProcessPath = function (pid, processHandle) {
+    var promise = fluid.promise();
+
+    // Check if the 'commandType' that is going to be executed is valid.
+    var validCommands = ["Open", "Close"];
```
If there's only ever going to be 2 choices, why not use a boolean and call something like `waitForClose` (or `waitForOpen`)?

(and rather than describing what `command` does, it will say what your code will do)

JavierJF replied:
Because calling it using a boolean flag like waitForClose seemed to me ambiguos in the false case, that would mean that:

* !waitForClose == waitForOpen *

I didn't considered that equality intuitive.

stegru replied:
Just 1 flag is needed:

```javascript
if (options.waitForClose) {
 // wait for close
} else {
 // wait for start
}
```
(or name it `waitForOpen` if the default of closing makes more sense)

JavierJF replied:
Yes yes, I knew, I wasn't just convinced that *the opposite* of waitForClose, being waitForOpen was natural, or evident. And that is why in first instance I chose strings. But it seems it looks evident to you, so I will go with the flag.

stegru commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -487,6 +350,460 @@ gpii.windows.getProcessPath = function (pid, processHandle) {
+    }
+
+    var keepTrying = function (rPromise, retries) {
+        if (retries <= 0) {
```
Why does it need to retry? If it failed the first time, why would it succeed the second?

JavierJF replied:
Because the command could fail in the first attempt, in the case of 'ClassicShell' for example we observed a delay of 6 seconds (depending on the system) while the daemon is not responsive to client petitions.

So a retry looked like the most logical thing to do in such cases when you can not expect responsiveness from external commands.

stegru commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -487,6 +350,460 @@ gpii.windows.getProcessPath = function (pid, processHandle) {
+                    isError: true,
+                    message: "Exceeded the number of retries trying to close process '" + processToClose + "'."
+                });
+            } else if (options.commandType === "Open") {
```
These 2 blocks are exactly the same, apart from 2 words.

JavierJF replied:
I can extract in small function and make the string completion with parameters if needed.

stegru commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -487,6 +350,460 @@ gpii.windows.getProcessPath = function (pid, processHandle) {
+            };
+
+            if (options.commandType === "Open") {
+                gpii.windows.waitForProcessStart(
```
These two blocks are exactly the same, apart from the function. I suggest doing something like:
```
var waitFor = <start> ? gpii.windows.waitForProcessStart : gpii...Termination;
waitFor(...);
```

JavierJF replied:
I like the idea, should be good to have a style-guide to decide this kind of situations for us, I use to trait function objects in a more classic way. I will try to minimize code with this kind of refactoring since now.

stegru replied:
Well, it's mainly for reducing duplicate code, rather than minimising it.

Another option is putting the return promise into a variable, and call .then on it once after.

stegru commented on gpii/node_modules/processHandling/processHandling.js:
```diff
@@ -487,6 +350,460 @@ gpii.windows.getProcessPath = function (pid, processHandle) {
+        }
+    };
+
+    keepTrying(promise, options.retries);
```
It should be ok to use `promise` and (a copy of) `options.retries` in the function.

JavierJF replied:
I know, that was just because I though it was more clear to follow it in that way.

stegru replied:
fine, I don't care.

stegru commented on gpii/node_modules/processHandling/test/testProcessHandling.js:
```diff
@@ -303,3 +320,403 @@ jqUnit.asyncTest("Testing getProcessPath", function () {
+            preLevels = [];
+
+            promise.resolve();
+            jqUnit.start();
```
put this in a `.then` at the end of closeAppWithMinForce

JavierJF replied:
Okay

stegru commented on gpii/node_modules/processHandling/test/testProcessHandling.js:
```diff
@@ -303,3 +320,403 @@ jqUnit.asyncTest("Testing getProcessPath", function () {
+
+        var promise = fluid.promise();
+
+        if (preLevels.length === 0) {
```
Couldn't these go into an array?
```
testData[preLevels.length].levelNum
testData[preLevels.length].preLevels
preLevels.push(preLevels.length+1)
```

They all seem to match length (or +1), so it might be nicer to do that instead.

JavierJF replied:
okay, I see what you mean. Yes it could avoid the repetition, but it was for a test, so I didn't care. I will change it.

amb26 replied:
:+1: Test cases are code too :)

JavierJF replied:
Haha sure! I'm working into all suggestions.

stegru commented on gpii/node_modules/WindowsUtilities/WindowsUtilities.js:
```diff
@@ -724,9 +848,16 @@ windows.accessFlag = function (object, flagName, setValue) {
 
-    for (var i = 0; i < array.length; ++i) {
-        ref.set(buf, i * size, array[i], windows.types[type]);
+    if (type === "TCHAR") {
```
In line 488 of SpiSettingsHandler, this function isn't called if is `TCHAR`.

JavierJF replied:
Removing it, was making checks for the function not taking into account the use case. Bad habit, we have talked about it before.

stegru commented on gpii/node_modules/WindowsUtilities/WindowsUtilities.js:
```diff
@@ -742,9 +873,16 @@ windows.arrayToBuffer = function (array, type) {
 
-    for (var i = 0; i < buffer.length / size; ++i) {
-        array[i] = ref.get(buffer, i * size, windows.types[type]);
+    if (type === "TCHAR") {
```
Like `arrayToBuffer`, this function isn't called if type is `TCHAR`.

See SpiSettingsHandler.js:83

JavierJF replied:
Yes, same as before.

stegru commented on gpii/node_modules/spiSettingsHandler/src/SpiSettingsHandler.js:
```diff
@@ -337,7 +482,11 @@ gpii.windows.spi.getPvParam = function (payload) {
 
     if (payload.options.pvParam.type === "array") {
-        pvParam = gpii.windows.arrayToBuffer(pvParam, payload.options.pvParam.valueType);
+        if (payload.options.pvParam.valueType === "TCHAR") {
```
An array of TCHARs is technically correct, but I think it might be better to have a new type of "string", and perform the UCS-2/UTF-8 conversion before and after the actual call to SystemParametersInfo

And look at `windows.stringToWideChar` and `windows.stringFromWideChar` for the actual conversion.

JavierJF replied:
Well, having that type will make things easier for the future for sure, in case we need to reuse it. I will look into it and comment the results to you.

stegru commented on gpii/node_modules/spiSettingsHandler/src/SpiSettingsHandler.js:
```diff
@@ -260,6 +315,96 @@ gpii.windows.spi.populateResults = function (payload, isNewValue, isGetting, res
+/**
+ * Checks and returns the value for the fWinIni parameter of the SystemParametersInfo function.
+ */
+gpii.windows.spi.getFWinIni = function (payload) {
```
This isn't called. Was this the duplicate of what

JavierJF replied:
You are right, was added but never removed.

stegru commented on gpii/node_modules/spiSettingsHandler/src/SpiSettingsHandler.js:
```diff
@@ -181,6 +197,32 @@ gpii.windows.spi.callProblematicSpi = function (pvParamType, action, uiParam, pv
     return promise;
 };
 
+gpii.windows.spi.checkWallpaperPath = function (path) {
```
`/** */`

stegru commented on gpii/node_modules/spiSettingsHandler/src/SpiSettingsHandler.js:
```diff
@@ -260,6 +315,96 @@ gpii.windows.spi.populateResults = function (payload, isNewValue, isGetting, res
+ */
+function splitInWords(input) {
+    var result = input.split(/[ ]+/);
+    if (result[(result.length - 1)] === "") {
```
`input.trim().split()`

stegru commented on gpii/node_modules/spiSettingsHandler/src/SpiSettingsHandler.js:
```diff
@@ -76,20 +79,31 @@ gpii.windows.spi.getCurrentSettings = function (payload) {
 
-gpii.windows.spi.systemParametersInfo = function (pvParamType, action, uiParam, pvParam) {
+gpii.windows.spi.systemParametersInfo = function (pvParamType, action, uiParam, pvParam, fWinIni) {
+    // Linter Workaround: Linter does not accept default parameters from ES6 spec.
```
So what?

stegru replied:
`fWinIni = fWinIni || 0` - but I don't think it's needed, ffi will assume undefined is 0.

stegru commented on provisioning/data/scripts/TabletView.ahk:
```diff
@@ -0,0 +1,23 @@
+SELECTED_MODE = %1%
+TABLETMODESTATE_DESKTOPMODE := 0x0
+TABLETMODESTATE_TABLETMODE := 0x1
+
```
Interesting hack, but I think C# is more appropriate.

JavierJF replied:
I also think it would be more appropriate, also, we have new tools that could give us much more better approaches than this, but the functionality was working and hasn't been touched since it was implemented. I will take a look to it and see how to create a better version for when this needs to reach master.

stegru commented on provisioning/data/scripts/ImmersiveShellRecover.ps1:
```diff
@@ -0,0 +1,13 @@
+<#
```
Do these scripts only get ran in our vagrant box? Or on the client?

My understanding of the `provisioning` directory is it contains scripts to provision our VM. They don't get put into the installer. Why not have a top-level directory called `data`?

JavierJF replied:
Yes, you are right, these scripts are copied during provisioning to the GPII data folder, which is the place we decided to keep for the scripts that will be later call in the system, if we don't have other working technique for doing the task.

stegru commented on provisioning/data/scripts/StartMenuRecover.ps1:
```diff
@@ -0,0 +1,49 @@
+  # Specify the current script path and name as a parameter
+  $newProcess.Arguments = $myInvocation.MyCommand.Definition;
+  # Indicate that the process should be elevated
+  $newProcess.Verb = "runas";
```
I know you know this: Either the UAC prompt will appear when keying in/out, or you'll have to disable the prompt. It's ok for a demo.

A better approach would be using this https://gist.github.com/stegru/7c8126259af943c4f8412710c2b53367 during install time (ran by the installer, as admin). This reduces the need to be administrator to change some registry values.

The best approach, however, is doing this kind of thing through a service (GPII-2338).

JavierJF replied:
Yes, this need to change. As you have probably saw, the rest of the script just changes some registry keys and restart a process, so probably this all needs to be rewrited when [this pull](https://github.com/GPII/windows/pull/120) enter master.

JavierJF committed:
GPII-2521: Changed function definition for making it more intuitive

JavierJF committed:
GPII-2521: Improved several functions stability.

    - Improved closeAppWithMinForce: Now it tries to manually restart the
      process, and wait for the application to shutdown. If that fails,
      then it changes to call closeApplications with "medium" and "high"
      force leves.
    - Improved process detection mechanims, detection by signals node api
      was error prone.

NOTE: Current impl needs to be refactored into one using processReporter

JavierJF committed:
GPII-2521: Added tests for new processHandling functions impl

JavierJF committed:
Merge branch 'master' of https://github.com/GPII/windows into GPII-2521

JavierJF committed:
GPII-2521: Fixed reference to current development universal

JavierJF committed:
GPII-2521: Added new settingsHandler for setting windows tablet mode

JavierJF committed:
GPII-2521: Added new settingsHandler for settings Windows 10 start menu layout

JavierJF committed:
GPII-2521: Added some convenient functions for processHandling

JavierJF committed:
Merge branch 'master' of https://github.com/GPII/windows into GPII-2521

JavierJF committed:
GPII-2521: Resolved forgotten merge conflicts

JavierJF committed:
GPII-2521: Removed check because rejecting promise 'halts' AcceptanceTests

JavierJF committed:
GPII-2521: Fixed linter issues

JavierJF committed:
GPII-2521: Added installation of required msvc components for native module
